modules/beta-private-cluster-update-variant/dns.tf

```
@@ -55,7 +51,7 @@ EOF
   }
 
   depends_on = [
-    null_resource.delete_default_kube_dns_configmap,
+    module.gcloud_delete_default_kube_dns_configmap,
```

Unfortunately it looks like tests are failing and I think the cause is Terraform not correctly traversing the dependency graph for modules. Instead I think we need to depend on the `wait` output.

```suggestion
    module.gcloud_delete_default_kube_dns_configmap.outputs.wait,
```

---

Fixed.

---

I ran `make docker_test_lint` locally and I don't think that works. Any ideas? I wish I could see the Cloud Build outputs...

```
Error: Invalid depends_on reference

  on ../../modules/beta-public-cluster/dns.tf line 111, in resource "kubernetes_config_map" "kube-dns-upstream-nameservers-and-stub-domains":
 111:     module.gcloud_delete_default_kube_dns_configmap.outputs.wait,

References in depends_on must be to a whole object (resource, etc), not to an
attribute of an object.
```

---

Hmm, you might have to incorporate it into an implicit dependency then. FYI Cloud Build is just running `make docker_test_integration` so you can verify locally.

---

`module.gcloud_delete_default_kube_dns_configmap.wait` might also work.

---

Still need to brainstorm how to add an implicit dependency, but it doesn't look like depending on `.wait` will work.

```
Error: Reference to undeclared output value

  on ../../dns.tf line 81, in resource "kubernetes_config_map" "kube-dns-upstream-namservers":
  81:     module.gcloud_delete_default_kube_dns_configmap.wait,

An output value with the name "wait" has not been declared in
module.gke.module.gcloud_delete_default_kube_dns_configmap.
```

---

This example from the gcloud module might help - https://github.com/terraform-google-modules/terraform-google-gcloud/tree/master/examples/dependency_example

---

Based on a little digging it appears that the `module_depends_on` "workaround" is only for when a module is depending on another module, not a resource depending on a module.

---

Yeah, I'm wondering if we could add an implicit dependency? Maybe via a bogus label on the resource?

---

Would you accept a PR that does this:

```
labels = {
  maintained_by = "terraform"
  terraform_depends_on = module.gcloud_delete_default_kube_dns_configmap.outputs.wait
}
```

---

I think that's a reasonable direction, though it might cause noisy diffs. How about `length(module.gcloud_delete_default_kube_dns_configmap.outputs.wait) > 1 ? "done" : "undone"`

---

Doesn't `wait >= 1` mean, yes wait some more?

---

No, the `wait` output is kind of just a dummy list of the inputs: https://github.com/terraform-google-modules/terraform-google-gcloud/blob/master/main.tf#L35

---

Ok sure, should it be greater than zero though?

---

Yeah that might work too.

---

Part of the issue might be that I have been testing this PR with version `0.3` of the gcloud module.

Could you please take a look at this issue: https://github.com/terraform-google-modules/terraform-google-gcloud/issues/19

---

I believe I cleared up the dependency issues in both `cluster.tf` and `dns.tf`. I also resolved the merge conflict.

---

The command `make docker_generate && make docker_generate_docs && make docker_test_lint` ran locally just fine.

---

Fixed.